### ISSUE TITLE
Adjustable cache adapter

### ISSUE THREAD

**Issue by Digi92** [mosparo/mosparo#284]
Hi,
we are now using mosparo in a Kubernetes environment and would like to be able to set an alternative caching method, as the current [FilesystemAdapter](https://github.com/mosparo/mosparo/blob/b9c79f93bee6a3fba928fb96465e1c445e303384/src/Helper/CleanupHelper.php#L31) could potentially cause problems with multiple containers.

If several containers are running simultaneously, the CleanUp job may be executed at the same time, triggered by the calls in the FrontendApiController.  (See [here](https://github.com/mosparo/mosparo/blob/b9c79f93bee6a3fba928fb96465e1c445e303384/src/Controller/Api/V1/Frontend/FrontendApiController.php#L91) and [here](https://github.com/mosparo/mosparo/blob/b9c79f93bee6a3fba928fb96465e1c445e303384/src/Controller/Api/V1/Frontend/FrontendApiController.php#L135)) 
An option to change the used cache adapter would probably be beneficial here.  
Such a configuration option would also be advantageous in the future, especially if caching is used in other places, too.  

In addition, an environment variable for deactivating the CleanUp calls in the FrontendApiController would be nice.
This would allow developer to better control when and how the CleanUp job is executed.

I have attached an example cache implementation that I have built to address this.
config/services.yaml
```
   mosparo.cache.service:
        class: Symfony\Contracts\Cache\CacheInterface
        factory: ['Mosparo\Factory\CacheFactory', 'getCache']
        arguments:
            $cacheAdapter: '%env(MOSPARO_CACHE_ADAPTER)%'
 
    Mosparo\Helper\CleanupHelper:
        arguments:
            $cache: '@mosparo.cache.service'
```

src/Factory/CacheFactory.php
```
<?php

namespace Mosparo\Factory;

use Symfony\Component\Cache\Adapter\AdapterInterface;
use Symfony\Component\Cache\Adapter\FilesystemAdapter;
use Symfony\Component\Cache\Adapter\RedisAdapter;

class CacheFactory
{
    public static function getCache(string $cacheAdapter = null): AdapterInterface
    {
        $cacheAdapter = $cacheAdapter ?? $_ENV['MOSPARO_CACHE_ADAPTER'] ?: 'filesystem';
        
        switch ($cacheAdapter) {
            case 'redis':
                $redisConnection = RedisAdapter::createConnection($_ENV['REDIS_URL'] ?? 'redis://localhost');
                return new RedisAdapter($redisConnection);
            case 'filesystem':
                return new FilesystemAdapter();
            default:
                throw new \InvalidArgumentException("Unsupported cache adapter type: $cacheAdapter");
        }
    }
}
```

src/Helper/CleanupHelper.php
```
...
protected CacheInterface $cache;

public function __construct(EntityManagerInterface $entityManager, ProjectHelper $projectHelper, LoggerInterface $logger, CacheInterface $cache)
     {
         $this->entityManager = $entityManager;
         $this->projectHelper = $projectHelper;
         $this->logger = $logger;
+         $this->cache = $cache;
     }
...
public function cleanup($maxIterations = 10, $force = false, $ignoreExceptions = true, $timeout = 1.5)
     {
-        $cache = new FilesystemAdapter();
-        $nextCleanup = $cache->getItem('mosparoNextCleanup');
-        $cleanupStartedAt = $cache->getItem('mosparoCleanupStartedAt');
+        $nextCleanup = $this->cache->getItem('mosparoNextCleanup');
+        $cleanupStartedAt = $this->cache->getItem('mosparoCleanupStartedAt');
 
```

.env
```
MOSPARO_CACHE_ADAPTER=redis
REDIS_URL=redis://localhost
```

What do you think of these points? 

Kind regards
Digi92

**Comment by zepich** [comment]
Hi @Digi92 

Thank you very much for your suggestion and the code you provided.

That's an excellent suggestion! I think it's essential not to use a file cache, especially for larger setups (multiple nodes) and/or failover setups, which would have to be synchronized between the nodes or cleared on all nodes. 

**Regarding the cleanup process in the frontend controller:**
I understand your request and agree with you in general. If you use the cronjob, you want to avoid doing any work in the FrontendController. But on the other side, one of the main priorities of mosparo is to automatically remove the data from the database to ensure that the data in mosparo are stored for only 14 days. The idea behind mosparo is not to store data forever (there are a lot of other good tools for this task).

I have considered this problem, and I think I have found a good solution. Instead of adding an environment variable to turn off the cleanup in the FrontendController, we add an environment variable to enable a 'grace period'. This grace period is set to one day. In the FrontendController, the cleanup is only executed if the time for the next cleanup is reached, plus the grace period. 

We schedule the next cleanup job in 6 hours. With the grace period, the FrontendController will not execute the cleanup until 24 hours after the planned job date and time. But suppose you regularly clean the database with the cronjob (or, in general, the cleanup command in the CLI). In that case, the end of the grace period (this 24 hours) is never reached because the cleanup command will schedule a new job, which means the grace period starts over again. But if you miss cleaning the database, the FrontendController will clean up the database when the grace period expires.

What do you think about this option? 

We will release the first beta of v1.3 today. Since these two changes do not change the UI, we will add them to the second beta in the next few days so we can release them in January with v1.3.

Thank you for your amazing work!

Kind regards,

zepich

**Comment by Digi92** [comment]
Hi @zepich,

i really like the idea of the grace period.  
This gives the admin more control over when the job is executed in order to utilise quieter times for the cleanup. The FrontendController then only intervenes in exceptional cases if the Cleanup via the cronjob does not run on time.

Kind regards,
DIgi92

**Comment by zepich** [comment]
Hi @Digi92 

As discussed, I've added the cache adapter and the grace period (see [9ef2a36](https://github.com/mosparo/mosparo/commit/9ef2a36becc6d9c5df61d83ac92a5c5091a56acc) and [f02d984](https://github.com/mosparo/mosparo/commit/f02d9847484c7f0c3e08bf9e912d0df056d223aa)).

**Cache adapter**
I've used your code mainly but made some adjustments. I've removed the use of `$_ENV` from the CacheFactory because, as far as I understand the Docker documentation about environment variables, it should work with the normal Symfony method to get the variables.

I've also added the Memcached adapter, so you can now choose between Redis, Memcached, and Filesystem.

**Grace period**
The grace period can be enabled by setting `MOSPARO_CLEANUP_GRACE_PERIOD_ENABLED` to `1`. If enabled (as discussed), the `FrontendController` will wait for an additional 24 hours to start the planned cleanup. If the `FrontendController` cannot finish the cleanup process and schedules a new cleanup in 10 minutes, the cleanup will be executed after 10 minutes without a grace period.

If the grace period is enabled and the cronjob runs once daily, then the FrontendController will never clean anything in the database (as you suggested).

Please let me know if you see any problems with my adjustments.

Thank you very much for your help!

Kind regards,

zepich

**Comment by Digi92** [comment]
Hi @zepich,
the changes look good. 

One suggestion for improvement: 
At [`$cleanupStart = $nextCleanup->get();`](https://github.com/mosparo/mosparo/blob/f02d9847484c7f0c3e08bf9e912d0df056d223aa/src/Helper/CleanupHelper.php#L43) a `clone` should be added, as otherwise `$cleanupStart->add(new DateInterval('PT24H'))` will change the value of `$nextCleanup` unintentionally.

Kind regards
Digi92

**Comment by zepich** [comment]
Hi @Digi92 

Perfect, thank you very much for your feedback!

You're right. Using a clone is safer and better. I've updated the code accordingly ([0f67bc9](https://github.com/mosparo/mosparo/commit/0f67bc9605c8e4c142ad8b629d84336d62c00704)).

Kind regards,

zepich

**Comment by zepich** [comment]
Added with v1.3

**Comment by zepich** [comment]
Hey @Digi92 

What do you think about storing this information regarding the cleanup in the database instead of a cache? A cache is temporary and can disappear, but the next cleanup timestamp is pretty important.

If you delete the cache, the cleanup process will be executed on the next frontend API request, which can then delay that request (so the work is done in the Frontend API instead of the cron job).

Or what do you think?

Kind regards,

zepich

**Comment by Digi92** [comment]
Hi @zepich ,
i see the advantage of storing the information in the database, especially when the cache is emptied by other processes.

One possible approach would be to extend the CacheFactory with a Symfony DB adapter. This would allow us to introduce a new service that is specifically responsible for managing the cleanup execution date.

However, I have concerns about performance. If a large number of frontend requests have to retrieve this information directly from the database instead of using a caching system like Redis, this could lead to a noticeable increase in load.

A possible compromise would be to continue using mosparo.cache.service as the primary source. If this is empty, a new mosparo.dbcache.service could retrieve the information from the database as a fallback. Only after a successful cleanup execution would we update the DB cache.

What do you think of this approach?

Kind regards
Digi92

**Comment by zepich** [comment]
Hi @Digi92 

Thank you very much for your response.

Honestly, I didn't have performance in my mind for this one. But you're right. It's an unnecessary load on the database.

The main thing I don't like (anymore) is that we store this important information in a cache. A cache is always dismissable, and we should not rely on it.

But I think I have a better solution.

1. We add a new entity, for example, `CleanupStatistic`.
2. After executing the cleanup, we store a new `CleanupStatistic` object with the date and time of the execution, how it was executed (CLI/API), how long it took, how much was deleted (if possible), and so on.
3. The cleanup statistic will be visible for the administrator in the admin area and/or as a chart (to prove that mosparo cleaned the database and to also check that the frontend API is not clearing the database, for example, if you want to see that).
4. When we need to know when we have to execute the next cleanup, we take the last stored `CleanupStatistic` object and add the required time (6 hours).
5. To increase performance, we store this calculated timestamp in the shared cache, which does not increase the load on the database.
6. If the cache is cleared (for whatever reason), we calculate the next execution timestamp again from the last stored `CleanupStatistic` object and store it again in the cache.

With that solution, we can solve multiple problems:

1. We never forget when we have to execute the next cleanup (it's always in the database).
2. We store the information in the shared cache for optimized performance so the database has no additional load.
3. We can get the information from the database if the shared cache **service** is unavailable (For example, Memcached/Redis service is down. Right now, this would generate an error 500 in the Frontend API).

What do you think?

Kind regards,

zepich

**Comment by Digi92** [comment]
Hi @zepich ,

**Comment by zepich** [comment]
Hi @Digi92 

Thank you very much for your feedback. Showing a hint in the submission list is a great idea!

I will create a new issue for developing this functionality for v1.4 and close this issue again.

Kind regards,

zepich
